Rewrite in Rust

ngryman opened the pull request.

PaperStrike commented:
WOW! This is really awesome! Sincere thanks for your contributions!

I haven't learned Rust at all so I only took some simple tests. It seems the computed sizes are increased by the changes, e.g., the represented size of a sample file [before](https://img.badgesize.io/https:/unpkg.com/@sliphua/pjax@2.4.0/dist/pjax.esm.min.js?compression=brotli), [after](https://img.runbots.io/https:/unpkg.com/@sliphua/pjax@2.4.0/dist/pjax.esm.min.js?compression=brotli).

ngryman commented:
@PaperStrike Thanks for testing this out!

I should have mentioned that but I changed Brotli's compression quality in the new version. It explains the differences.

Node.js defaults it to the maximum value (`11`). It produces a smaller file at the expense of much more CPU cycles. That's why I think servers out there have lower defaults. For example, Apache sets it to `5` and Nginx is `6` by default. The reason is that the additional compression time would likely be slower than transferring a slightly bigger file, which would defeat the purpose of speeding transfers.

If we take `unpkg` as an example. By inspecting the transferred size with Chrome DevTools, we can see that it takes `4.1kb` and not `3.45kb` as advertised by the older version. So it's likely `unpkg` uses something close to the configuration of the new `badgesize` version.

After your comment, I dug a little deeper, and I adjusted Brotli's options a little bit to try to match a reasonable default configuration (cf. c4a934d). The reported size of this new version is still above the older (eg. `3.9kb` for your library), but I think it's for the best. I believe it's better to advertise something close to real-world sizes rather than optimal sizes.

styfle commented:
@ngryman I realize you've already done a lot of work on the Rust implementation here but I wanted to reach out because the  $60/month for Vercel doesn't sound right. The Pro plan is $20/month and should offer plenty of bandwidth and function execution without hitting any overages. 

That being said, Vercel offers sponsorships to cover the cost of open source projects so you don't need to worry about billing at all 😃 

Take a look at https://vercel.com/support/articles/can-vercel-sponsor-my-open-source-project

ngryman commented:
Hey @styfle, thanks for the heads up, I appreciate it.

I already have the pro plan, but unfortunately, I'm reaching the Execution quota. I'm consuming 2 × 100 GB-Hours which cost me an additional $40.

For the sponsorship, I already asked for a sponsorship last year but got the following response:

```
At the moment, we are sponsoring OSS packages that fit the criteria here.
Serverless Functions are indeed a premium offering in our platform and that is why we are not prepared
to sponsor projects heavily using this type of resource.
If you don't want to use Vercel to deploy this service, perhaps a serverful environment may suit your needs.
```

rauchg commented:
@ngryman definitely happy to help out out here

PaperStrike commented:
@ngryman Yes, I agree real-world sizes are better.

The transferred size shown in Chrome DevTools contains everything in the response including the headers, which will increase the size by around 0.4kB. You can take that into account to be more accurate.-)

ngryman commented:
@PaperStrike Oh good catch for the headers. My example with `unpkg` is probably not the best then 😅 
I'll keep the current settings for now and if folks complain about it, I'll revisit it.

ngryman commented:
@rauchg Oh cool 🎉  I guess I'll ask directly, could this project be eligible for your OSS sponsorship? If so, I can port this PR to use the Vercel Rust runtime.

styfle commented:
@ngryman Yes, I confirmed with @rauchg that Vercel will sponsor this project 👍 

Once you add [a logo and link](https://vercel.com/support/articles/can-vercel-sponsor-my-open-source-project) to your project, contact support and we'll apply the sponsorship to your project! 🎉 

I hope that helps you ship faster, thanks! 🤗

ngryman commented:
Oh wow, thank you so much guys! That's just awesome 🎉
